[tuya] Initial contribution (#213)
smarthomej/addons: bundles/org.smarthomej.binding.tuya/README.md

`project` things must be configured manually and are needed for discovery only.

`tuyadevice` things represent a single device.
They can be configured manually or by discovery.

## Discovery

Discovery is supported for `tuyadevice` things.
By using discovery all necessary setting of the device are retrieved from your cloud account.

## Thing Configuration

### `project`

First create and link a Tuya Develop Account:

- Go to `iot.tuya.com` (the Tuya developer portal) and create an account. 
You can choose any credentials (email/password) you like (it is not necessary that they are the same as in the app). 
After confirming your account, log in to your new account. 
- On the left navigation bar, select "Cloud", then "Create new Cloud project" (upper right corner).
Enter a name (e.g. "My Smarthome"), select "Smart Home" for "Industry" and "Development Method".
For security reasons, select only the "Data Center" that your app is connected to (you can change that later if you select the wrong one). 
Select "IoT Core", "Authorization" and "Device Status Notification" as APIs.
- You should be redirected to the "Overview" tab of your project. 

> Write down (or copy) "Access ID/Client ID" and "Access Secret/Client Secret" (ypu can always look it up in your account).

Author: ```suggestion
Write down (or copy) "Access ID/Client ID" and "Access Secret/Client Secret" (you can always look it up in your account).
```

- In the upper menu bar, select the "Devices" tab, then go to "Link Tuya App Account" and link you App account.

> Add a `project` and enter your credentials (`username`/`password`, from the app - not your cloud account!) and the cloud project credentials (`accessId`/`accessSecret`).

Author: ```suggestion

The next steps are performed in openHAB's Main UI:

Add a `project` and enter your credentials (`username`/`password`, from the app - not your cloud account!) and the cloud project credentials (`accessId`/`accessSecret`).
```

The `countryCode` is the international dial prefix of the country you registered your app in (e.g. `49` for Germany or `43` for Austria).
Depending on the app you use, set `schema` to `tuyaSmart` (for the Tuya Smart app) or `smartLife` (for the Smart Life app).
The `datacenter` needs to be set to the same value as in your IoT project.

The thing should come online immediately.

If the thing does not come online, check 

> - if you entered the correct country code (check in the App if you accidentally choose a wrong country)

Author: ```suggestion
- if you really used the app and not the developer portal credentials
- if you entered the correct country code (check in the App if you accidentally choose a wrong country)
```

- check if you selected the correct "Data Center" in your cloud project (you can select more than one for testing).

### `tuyaDevice`

The best way to configure a `tuyaDevice` is using the discovery service.

The mandatory parameters are `deviceId`, `productId` and `localKey`.
The `deviceId` is used to identify the device, the `productId` identifies the type of the device and the `localKey` is a kind of password for access control.
These parameters are set during discovery.
If you want to manually configure the device, you can also read those values from the cloud project above.

For line powered device on the same subnet `ip` address and `protocol` version are automatically detected.
Tuya devices announce their presence via UDP broadcast packets, which is usually not available in other subnets.
Battery powered devices do not announce their presence at all.
There is no clear rule how to determine if a device has protocol 3.3 or 3.1.
It is recommended to start with 3.3 and watch the log file if it that works and use 3.1 otherwise.

> *Note:* Support for protocol 3.1 is considered experimental.

Author: ```suggestion
*Note:* Support for protocol 3.1 is considered experimental.
In case something is not working, please open an issue on [GitHub](https://github.com/smarthomej/addons/issues) and add TRACE level logs.
```

## Channels

> For auto-discovered devices the schema of the device is requested from the cloud.
> Based on that channels are added to the thing on first startup.
> 
> In case you manually configured the device (or no schema was retrieved during discovery), a database is used to lookup the correct schema.
> The device will change to OFFLINE status if this is not successful.

Author: ```suggestion
Channels are added automatically based on device schemas on first startup.
The binding first tries to get it from a database of known device schemas.
If no schema is found a schema retrieved from the cloud during discovery is used (if applicable).

The device will change to OFFLINE status if no device schema could be determined.
```
